test/csuite/wt12015_backup_corruption/main.c

```diff
@@ -0,0 +1,442 @@
+    testutil_check(cursor->close(cursor));
+
+    /* Cleanup. */
+    testutil_check(session->close(session, NULL));
+    testutil_check(conn->close(conn, NULL));
+}
+
+/*
+ * do_work_before_failure --
+ *     Do a bunch of work, including taking backups and checkpoints, before we inject the failure.
+ */
+static void
+do_work_before_failure(WT_CONNECTION *conn, WT_SESSION *session)
+{
+    int i;
+    char backup_home[64], backup_id[32], src_backup_home[64], src_backup_id[32];
+
+    for (i = 0; i < NUM_BACKUPS; i++) {
+        populate_table(session, TABLE_URI, (uint32_t)i, 100 * WT_THOUSAND);
+        testutil_check(session->checkpoint(session, NULL));
+        populate_table(session, TABLE_URI, (uint32_t)i + 1, 100 * WT_THOUSAND);
+        testutil_check(session->checkpoint(session, NULL));
+        populate_table(session, TABLE_URI, (uint32_t)i + 2, 100 * WT_THOUSAND);
+
+        testutil_snprintf(backup_home, sizeof(backup_home), BACKUP_BASE "%d", i);
+        testutil_snprintf(backup_id, sizeof(backup_id), "ID%d", i);
+        if (i == 0) {
+            printf("Create full backup %d\n", i);
+            testutil_backup_create_full(conn, WT_HOME_DIR, backup_home, backup_id, true, 32, NULL);
+        } else {
+            printf("Create incremental backup %d from %d\n", i, i - 1);
+            testutil_snprintf(src_backup_home, sizeof(src_backup_home), BACKUP_BASE "%d", i - 1);
+            testutil_snprintf(src_backup_id, sizeof(src_backup_id), "ID%d", i - 1);
+            testutil_backup_create_incremental(conn, WT_HOME_DIR, backup_home, backup_id,
+              src_backup_home, src_backup_id, false /* verbose */, NULL, NULL, NULL);
+        }
+    }
+}
+
+/*
+ * do_work_after_failure --
+ *     Do work after an injected failure: Reopen the database, do more work, create another
+ *     incremental backup, and verify it.
+ */
+static void
+do_work_after_failure(bool backup_from_min)
+{
+    WT_CONNECTION *conn;
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    WT_SESSION *session;
+    int i, id;
+    char backup_home[64], backup_id[32], src_backup_home[64], src_backup_id[32], *str;
+
+    /* Reopen the database and find available backup IDs. */
+    testutil_wiredtiger_open(opts, WT_HOME_DIR, env_config, NULL, &conn, false, false);
+    testutil_check(conn->open_session(conn, NULL, NULL, &session));
+
+    ret = session->open_cursor(session, "backup:query_id", NULL, NULL, &cursor);
+    testutil_check(ret);
+    id = -1;
+    while ((ret = cursor->next(cursor)) == 0) {
+        testutil_check(cursor->get_key(cursor, &str));
+        testutil_assert(strncmp(str, "ID", 2) == 0);
+        i = atoi(str + 2);
+        if (id < 0)
+            id = i;
+        else if (backup_from_min)
+            id = WT_MIN(id, i);
+        else
+            id = WT_MAX(id, i);
+        printf("Found backup %d\n", i);
+    }
+    testutil_assert(ret == WT_NOTFOUND);
+    testutil_check(cursor->close(cursor));
+    testutil_assert(id >= 0);
```

Comment:
I don't think there is quite anything else that we need...

I guess we could add assertions for making sure that we get the backup IDs that we expect if the code runs correctly? I didn't want add that because the correct answer depends on the solution to the problem—i.e., whether we log and replay changes to backup IDs vs. make turtle file update atomic with the metadata table update. And even then, the correct answer will change depending on whether logging is enabled in the test.